### ISSUE TITLE
InstaStack - compliance check

### ISSUE THREAD

**Issue by bluecypher** [ONDC-Official/v1.1.0-logs#206]
## Flow 1

- /context/ttl is not required for /on_action APIs

**/on_search**
- Either one of fixed (range) or split (frequency and times) timings should be provided in /bpp/providers/locations/time
- @ondc/org/contact_details_consumer_care should be in the format "name,email,contactno" in /bpp/providers/items
- category in serviceability construct should be defined only for category ids in bpp/providers/items/category_id

**/on_select**
- @ondc/org/TAT (O2D) can't be equal to @ondc/org/time_to_ship (O2S) in /on_search

**/on_init**
- /billing/address must have required property 'locality'
- /fulfillments/end/location/address must have required property 'locality'
- /payment/@ondc/org/buyer_app_finder_fee_type should be 'percent' (with lowercase "p")
- many invalid attributes in payment object (type, collected_by, @ondc/org/withholding_amount, @ondc/org/return_window, @ondc/org/settlement_basis, @ondc/org/settlement_window)
- /payment/@ondc/org/settlement_details/settlement_type must be equal to one of the allowed values (upi,neft,rtgs)
- buyer app finder fee amount can't change in /on_init
- item.quantity.available and item.quantity.maximum not required in quote.breakup

**/on_confirm**
- order.created_at timestamp cannot change in /on_confirm
- item.quantity.available and item.quantity.maximum not required in quote.breakup

**/on_status**
- 'tags' is not mandatory in /items for 'Order-delivered'
- /documents/label must be 'Invoice'
- pickup(fulfillments/start/time/timestamp) and delivery(fulfillments/end/time/timestamp) time is required when fulfillment state is Order-delivered

**/on_update**
- Order should have required properties 'payment', 'created_at', 'updated_at'.
- ReverseQC must have required property 'end'
- ReverseQC should have pickup timestamp (start/time/timestamp) when the fulfillment state is 'Order-picked-up'
- Message Ids for /update and /on_update APIs should be same
- ReverseQC pickup (/start/time/timestmap) and delivery time (end/time/timestamp) cannot be same when fulfillment state is 'Order-delivered'
- message_id for all the unsolicited /on_update should be same for a particular return request (/update)

## Flow 2

- similar issues as Flow 1

**/on_select** (non serviceable)

- Error object (domain error and code) required when the fulfillment is non-serviceable

**/on_cancel**

- context/transaction_id must not change

**/on_status** (post cancel)

- context/transaction_id must not change
- context/timestamp for /on_status_post_cancel can't be before the timestamp for order cancellation in /on_cancel
- cancellation_reason_id missing in tags

## Flow 3

- similar issues as Flow 1

## Flow 4

- similar issues as Flow 1

**/on_select**

- Error object (domain error and code) is required when the quantity is not in stock
- message_id for /select and /on_select does not match

## Flow 5

- similar issues as Flow 1

@adil-instastack

**Comment by adil-instastack** [comment]
Hi Sandeep,

Feedback has been implemented. Commit is available on

https://github.com/ONDC-Official/v1.1.0-logs/pull/212

Please review.
Thanks in advance.

**Comment by adil-instastack** [comment]
Hi @Sandeep Shahi ***@***.***> ,

Did you get a chance to review our logs?
Thanks in advance.

**Comment by bluecypher** [comment]
@adil-instastack, the recent PR created conflict, please resubmit the logs.

**Comment by adil-instastack** [comment]
https://github.com/ONDC-Official/v1.1.0-logs/pull/219

Done

**Comment by abhinavv245** [comment]
## Flow 1

**/on_init**

- Billing object mismatches in /init and /on_init
(address.door is missing in /on_init)
- Discrepancies between the quote object in /on_select and /on_init (packing charges, tax, convenience fee missing in /on_select)

**/on_confirm**

- Timestamp for /confirm api cannot be equal to /on_confirm api
- 'tags' is not mandatory in /items (should be used for return status)

**/on_status (unsolicited-delivered)**

- delivery time (/fulfillments/end/time/timestamp) should be equal to context.timestamp when the fulfillment state changes to 'Order-delivered'

**/on_update**

- order.updated_at timestamp should be updated as per context.timestamp

**/on_update (return picked up)**

- pickup time(fulfillments/start/time/timestamp) for ReverseQC fulfillment should match the context.timestamp

**/on_update (return delivered)**

- delivery time (fulfillments/end/time/timestamp) for ReverseQC fulfillment should match the context.timestamp

## Flow 2

- similar issues as Flow 1

**/on_select (non-serviceable)**

- 'type' is missing in Error object (DOMAIN-ERROR)
- delivery line item should not be present in case of non-serviceability

**on_status (post cancel)**

- unsolicited on_status (post cancel) is not required
- context.transaction_id must not change
- fulfillment state should change to 'Cancelled'
- order.updated_at should be equal to /on_cancel context.timestamp

## Flow 3 

- similar issues as Flow 1

## Flow 4

- similar issues as Flow 1

**/on_select**
- type' is missing in Error object (DOMAIN-ERROR)
- item id:108668 cannot be unavailable when inventory is available (quantity.available.count is more than the required quantity)

## Flow 5

- similar issues as Flow 1

@adil-instastack

**Comment by adil-instastack** [comment]
Hi Abhinav,

Feedback has been implemented.

New pull request is at
https://github.com/ONDC-Official/v1.1.0-logs/pull/229

Kindly review. Hope we have completed all required changes this time.
Thanks a lot in advance.

**Comment by abhinavv245** [comment]
## Flow 1

-  order.updated_at timestamp should be updated as per the context.timestamp 

**/on_select**

- '@ondc/org/TAT' for Immediate delivery can be 60 minutes (max)
- In Error object, 'type' should be 'DOMAIN-ERROR' (instead of 'seller app')

## Flow 2, Flow 3, Flow 4, Flow 5

- similar issues as Flow 1

@adil-instastack

**Comment by adil-instastack** [comment]
Hi Abhinav,

Thanks for the quick feedback, much appreciated.

Fixes on_select have been made and I will redeploy logs.

However, your first feedback seems to be already implemented.
order.updated_at timestamp matches context.timestamp.
Attaching relevant 3 payloads for reference.

Can you let me know if I am missing something?

**Comment by adil-instastack** [comment]
Hi Abhinav,

Can you please help us on the timestamp query below.
Would like to resubmit logs asap.

**Comment by abhinavv245** [comment]
Hi @adil-instastack ,
As discussed, order.updated_at feedback needs to be implemented. 
Please resubmit the logs after making the required changes.
Thanks.

**Comment by adil-instastack** [comment]
Hi Abhinav,

Thanks a lot for the clarifications.

Have redeployed the logs.
Here is the pull request.

https://github.com/ONDC-Official/v1.1.0-logs/pull/232

Hope all is done now :)

**Comment by adil-instastack** [comment]
Hi Abhinav,

Had to redeploy due to conflict.

Here is the new pull request
https://github.com/ONDC-Official/v1.1.0-logs/pull/234

Please help us close this today if possible.

**Comment by abhinavv245** [comment]
## All Flows

**/on_search**

- /message/catalog/bpp/providers/items/@ondc/org/statutory_reqs_prepackaged_food must have required property 'imported_product_country_of_origin' as per advisory issued (updated in v1.1.0 API Contract)

**Comment by adil-instastack** [comment]
Okay.. will update thanks.

**Comment by adil-instastack** [comment]
Do I need to submit all the logs again :(

**Comment by adil-instastack** [comment]
Hi Abhinav,

Have added country of origin on_search.
New pull request is at

https://github.com/ONDC-Official/v1.1.0-logs/pull/238

Kindly review.
Thanks in advance.

**Comment by bluecypher** [comment]
## Flow 1

**/on_search**

- /message/catalog/bpp/providers/0/items/0/@ondc/org/statutory_reqs_prepackaged_food/imported_product_country_of_origin country of origin must be in ISO 3166-1 alpha-3 code format (should be IND instead of INDIA)
- Follow [enhanced sub-category list](https://docs.google.com/spreadsheets/d/1ayRbp-WmXwwbzp7z1MgRO0NuKZM1AQk4GGZ8SE4NTnw/edit#gid=0) for the correct mapping of category_id (@ondc/org/mandatory_reqs_veggies_fruits is required only in case of "Fruits and Vegetables")
- bpp/providers/fulfillments "type" and "id" are invalid attributes

**/on_select**

- tax, packing and misc in quote/breakup should not be included if 0

**/on_confirm**

- the /on_confirm response took more than 5 seconds, reason for the delay?
- Ensure that the /confirm and /on_confirm flow follows the specifications outlined in the API contract
- The contact details in "fulfillments/end/contact/" should match those in the /confirm response.

**/on_update (Return Approved)**

- took more than context/ttl time to respond to /update (will be considered as a stale response)
- Is the return request set for auto approval?
- ReverseQC fulfillment in /fulfillments[1] should not be directly copied from the contract but rather be aligned with the current transaction

## Flow 2

**on_cancel (unsolicited)**

- use valid seller side cancellation reason code as [documented](https://docs.google.com/document/d/1M-lZSduYMFKIk1V6d8QLt-j-16-rVzYVdPn0pmbkclk/edit#heading=h.g7gnblpivbmp)

## Flow 3, 4, 5

- Apply the necessary corrections mentioned in Flow 1.


@adil-instastack

**Comment by adil-instastack** [comment]
Hi Sandeep,

Have made the required changes.

Pull request is at https://github.com/ONDC-Official/v1.1.0-logs/pull/243

Also, please find my comments below.

Flow 1

*/on_search*

   - ***@***.***/org/statutory_reqs_prepackaged_food/imported_product_country_of_origin
   country of origin must be in ISO 3166-1 alpha-3 code format (should be IND
   instead of INDIA) *- DONE*
   - Follow enhanced sub-category list
   <https://docs.google.com/spreadsheets/d/1ayRbp-WmXwwbzp7z1MgRO0NuKZM1AQk4GGZ8SE4NTnw/edit#gid=0>
for
   the correct mapping of category_id ***@***.***/org/mandatory_reqs_veggies_fruits
   is required only in case of "Fruits and Vegetables") *- DONE*
   - bpp/providers/fulfillments "type" and "id" are invalid attributes *- *
   *DONE*

*/on_select*

   - tax, packing and misc in quote/breakup should not be included if 0 *-
   DONE*

*/on_confirm*

   - the /on_confirm response took more than 5 seconds, reason for the
   delay? *- Have optimized, it's < 1 sec now.*
   - Ensure that the /confirm and /on_confirm flow follows the
   specifications outlined in the API contract *- It does, let me know if
   there are any specific discrepancies. *
   - The contact details in "fulfillments/end/contact/" should match those
   in the /confirm response. *- DONE*

*/on_update (Return Approved)*

   - took more than context/ttl time to respond to /update (will be
   considered as a stale response) *- Actually there was one more state,
   Return_Initiated which was implemented but i had not shared the log
   earlier. Have added this now. All responses are returned generally within
   0.5 secs.*
   - Is the return request set for auto approval? *- Yes, on
   staging/pre-prod for easier testing. This would be switched to manual
   approval on prod.*
   - ReverseQC fulfillment in /fulfillments[1] should not be directly
   copied from the contract but rather be aligned with the current
transaction *-
   DONE. Fulfillment details are as per the current order.*

Flow 2

*on_cancel (unsolicited)*

   - use valid seller side cancellation reason code as documented
   <https://docs.google.com/document/d/1M-lZSduYMFKIk1V6d8QLt-j-16-rVzYVdPn0pmbkclk/edit#heading=h.g7gnblpivbmp>
    *- DONE*

Flow 3, 4, 5

   - Apply the necessary corrections mentioned in Flow 1. *- DONE*

**Comment by adil-instastack** [comment]
Hi Sandeep,

Did you get a chance to review our logs.

Thanks in advance.

**Comment by bluecypher** [comment]
**/on_search**

- /message/catalog/bpp/providers/items veg/non-veg categorization is mandatory for F&B categories or packaged foods in /tags
- /message/catalog/bpp/providers/items/descriptor/images are optional for F&B categories and buyer apps don't process more than 3.

**/on_confirm**

- why is store name changed in /fulfillments/start/location/descriptor/name?
- instructions are optional (used for providing PCC and DCC)
- What is the reason for the extended estimated pickup and delivery time for "Immediate Delivery"?
- invoice should be provided only after shipment

**/on_update**

- Does the seller app receive the "/update" (FYI call) for the refund? If it does, how is it processed?

## Flow 2

**/on_cancel (unsolicited)**

- context/city code should not change
- why is there a change in txn id? (even though the timestamp and order ID indicate the same transaction)

## Flow 4

**/on_search**

- the response must be triggered within 5 sec of /search call

**/on_select**

- why is quotation provided for 10 quantity of item: "3350" even though only 2 were selected?
- flow must cover item unavailability scenario


@adil-instastack

**Comment by adil-instastack** [comment]
Hi Sandeep,

Thanks for the feedback.
Have made the required changes. Pull request is at
https://github.com/ONDC-Official/v1.1.0-logs/pull/258

Also, pls find my comments below.


*/on_search*

   - /message/catalog/bpp/providers/items veg/non-veg categorization is
   mandatory for F&B categories or packaged foods in /tags *- done. have
   added tags for veg/non-veg*
   - /message/catalog/bpp/providers/items/descriptor/images are optional
   for F&B categories and buyer apps don't process more than 3. - *done.
   will return a max of 3 images *

*/on_confirm*

   - why is store name changed in
   /fulfillments/start/location/descriptor/name? *- fixed*
   - instructions are optional (used for providing PCC and DCC) *- removed.
   will add as required *
   - What is the reason for the extended estimated pickup and delivery time
   for "Immediate Delivery"? -* Incorrect mapping. Have changed to "Same
   Day Delivery" now*
   - invoice should be provided only after shipment - *done. invoice link
   will be available when order is shipped*

*/on_update*

   - Does the seller app receive the "/update" (FYI call) for the refund?
   If it does, how is it processed?
   - * yes. we do two things. 1. append the object provided by the buyer to
   payment/settlement_details array. 2. we update the order updated_at
   timestamp. We use the settlement array for payment recon. Have added 2 more
   log files to case 1 which shows this request and response.*

Flow 2

*/on_cancel (unsolicited)*

   - context/city code should not change - *fixed*
   - why is there a change in txn id? (even though the timestamp and order
   ID indicate the same transaction) - *fixed*

Flow 4

*/on_search*

   - the response must be triggered within 5 sec of /search call -* fixed. *

*/on_select*

   - why is quotation provided for 10 quantity of item: "3350" even though
   only 2 were selected? *- fixed*
   - flow must cover item unavailability scenario

**Comment by bluecypher** [comment]
**/on_status_delivered**

- unsolicited on_status for previous fulfillment states should be provided for verification
- /payment/params must have required property 'transaction_id' and payment/status should be "PAID"

**/on_update**

- message_id should be the same as in /update (should not match with /confirm)
- /on_update_billing: invalid response of /update_billing. Why is this call initiated?
- /on_update_return_delivered: why are there 2 objects for same refund settlement trail


@adil-instastack

**Comment by adil-instastack** [comment]
Hi Sandeep,

Feedback implemented.
New pull request is at https://github.com/ONDC-Official/v1.1.0-logs/pull/269
Kindly review.

Also, please find my comments below.

*/on_status_delivered*

   - unsolicited on_status for previous fulfillment states should be
   provided for verification *- added*
   - /payment/params must have required property 'transaction_id' and
   payment/status should be "PAID" *- done, only prepaid orders have been
   submitted.*

*/on_update*

   - message_id should be the same as in /update (should not match with
   /confirm) -* fixed*
   - /on_update_billing: invalid response of /update_billing. Why is this
   call initiated? -* we thought we should send back on_update response
   when we receive the refund object. Have suppressed this.*
   - /on_update_return_delivered: why are there 2 objects for same refund
   settlement trail *- fixe*d

**Comment by bluecypher** [comment]
**/on_status**

- items/tags are not required (only for part returned/cancelled items)
- order pickup time is not captured correctly (should be from Order-picked-up state)


@adil-instastack

**Comment by adil-instastack** [comment]
Hi Sandeep,

Can you please further elaborate on the below feedback.

   - order pickup time is not captured correctly (should be from
   Order-picked-up state)

Not sure what we are doing wrong.

**Comment by bluecypher** [comment]
@adil-instastack, The logs do not include the interim state "Order-picked-up," which should be utilized to determine the actual pickup time in the "/start/time/timestamp" field instead of the "Out-for-delivery" state.

**Comment by adil-instastack** [comment]
Thanks Sandeep for the clarification.

Fixes have been made.
New pull request is at https://github.com/ONDC-Official/v1.1.0-logs/pull/279

Kindly review at your earliest convenience.

**Comment by adil-instastack** [comment]
Hi Sandeep,

Did you get a chance to review our logs yet?

We have completed all of the other requirements and are just waiting
go-ahead on logs to go live.
Would appreciate if you can help us go live sooner :)

**Comment by bluecypher** [comment]
**/on_search**

- /message/catalog/bpp/providers/0/items/0 veg/non-veg categorization is mandatory for F&B categories or packaged foods in /tags. Why is this removed again?
- /message/catalog/bpp/providers/0/items/1/descriptor/images must NOT have more than 3 images.
- store enable/disable timestamp (/bpp/providers/time/timestamp) must match context.timestamp
- /bpp/providers[1]/locations[0]/gps coordinates must be specified with at least six decimal places of precision.

**/on_init**

- item/quantity is optional in quote/breakup in /on_init and thereafter


@adil-instastack, Why do these issues resurface despite already being resolved on your end?

**Comment by adil-instastack** [comment]
Hi Sandeep,

We had made some db optimisations which caused some breaking changes.

Will fix and validate with tool before resubmitting.

**Comment by adil-instastack** [comment]
Hi Sandeep,

Have resubmitted after validating with log-validator tool.
log_report.md added for each flow in the commit.

New pull request is at:
https://github.com/ONDC-Official/v1.1.0-logs/pull/286

Please review. Trust we have fixed all issues now.

**Comment by bluecypher** [comment]
@adil-instastack, fulfillment id is "1" and reverse QC fulfillment is "REVERSE-QC", what happens in case of multiple fulfillment and multiple returns?

**Comment by adil-instastack** [comment]
Hi Sandeep,

The fulfilment IDs are dynamic.

In case of multiple fulfilment or return it will assign 2 or REVERSE-QC2
respectively.

Let me know in case you want me to submit logs for this case.

**Comment by bluecypher** [comment]
@BLR-0118, the Instastack Seller App (v1.1.0) logs seem fine. Please proceed with your review.

**Comment by adil-instastack** [comment]
Hi Supriyo,

Did you get a chance to review our logs.


Thanks & Regards,
Ajinkya Shelar
